src/resolver.rs

```diff
@@ -0,0 +1,385 @@
+        } else {
+            self.errors.push(SourceError {
+                message: "variable not found".to_string(),
+                node_id: unbound_node_id,
+                severity: Severity::Error,
+            })
+        }
+    }
+
+    pub fn resolve_block(
+        &mut self,
+        node_id: NodeId,
+        block_id: BlockId,
+        reused_scope: Option<ScopeId>,
+    ) {
+        let block = self
+            .compiler
+            .blocks
+            .get(block_id.0)
+            .expect("internal error: missing block");
+
+        if let Some(scope_id) = reused_scope {
+            self.enter_existing_scope(scope_id);
+        } else {
+            self.enter_scope(node_id);
+        }
+
+        for inner_node_id in &block.nodes {
+            self.resolve_node(*inner_node_id);
+        }
+        self.exit_scope();
+    }
+
+    /// Enter an existing scope frame, e.g., a block or a closure
+    pub fn enter_existing_scope(&mut self, scope_id: ScopeId) {
+        self.scope_stack.push(scope_id);
+    }
+
+    /// Enter a new scope frame, e.g., a block or a closure
+    pub fn enter_scope(&mut self, node_id: NodeId) {
+        self.scope.push(Frame::new(FrameType::Scope, node_id));
+        self.scope_stack.push(ScopeId(self.scope.len() - 1));
+    }
+
+    /// Exit a scope frame, e.g., when reaching the end of a block or a closure
+    ///
+    /// When exiting a scope frame, all overlays (and corresponding light frames) are removed along
+    /// with the removed frame.
+    pub fn exit_scope(&mut self) -> ScopeId {
+        match self
+            .scope_stack
+            .iter()
+            .rposition(|scope_id| self.scope[scope_id.0].frame_type == FrameType::Scope)
+        {
+            None => panic!("internal error: no scope frame to exit"),
+            Some(pos) => {
+                let scope_id = self.scope_stack[pos];
+                self.scope_stack.truncate(pos);
+                scope_id
+            }
+        }
+    }
+
+    pub fn define_variable(&mut self, var_name_id: NodeId, is_mutable: bool) {
+        let var_name = self.compiler.get_span_contents(var_name_id);
+        let var_name = trim_var_name(var_name).to_vec();
+
+        let current_scope_id = self
+            .scope_stack
+            .last()
+            .expect("internal error: missing scope frame id");
+
+        self.scope[current_scope_id.0]
+            .variables
+            .insert(var_name, var_name_id);
+
+        let var = Variable { is_mutable };
+        self.variables.push(var);
+        let var_id = VarId(self.variables.len() - 1);
+
+        // let the definition of a variable also count as its use
+        self.var_resolution.insert(var_name_id, var_id);
+    }
+
+    pub fn find_variable(&self, var_name: &[u8]) -> Option<NodeId> {
+        for scope_id in self.scope_stack.iter().rev() {
+            if let Some(id) = self.scope[scope_id.0]
+                .variables
+                .get(trim_var_name(var_name))
```

Comment:
Rather than having to trim it all the time, you could trim when the variable is parsed and then everywhere else assume it's trimmed.